src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -223,66 +281,41 @@ static ucs_status_t uct_cuda_ipc_rkey_release(uct_component_t *component,
 }
 
 static ucs_status_t
-uct_cuda_ipc_mem_reg_internal(uct_md_h uct_md, void *addr, size_t length,
-                              unsigned flags, uct_cuda_ipc_key_t *key)
+uct_cuda_ipc_mem_reg(uct_md_h md, void *address, size_t length,
+                     const uct_md_mem_reg_params_t *params, uct_mem_h *memh_p)
 {
-    ucs_log_level_t log_level;
+    uct_cuda_ipc_memh_t *memh;
     CUdevice cu_device;
-    ucs_status_t status;
-
-    ucs_assert((addr != NULL) && (length != 0));
-
-    log_level = (flags & UCT_MD_MEM_FLAG_HIDE_ERRORS) ? UCS_LOG_LEVEL_DEBUG :
-                UCS_LOG_LEVEL_ERROR;
-
-    status    = UCT_CUDADRV_FUNC(cuIpcGetMemHandle(&key->ph, (CUdeviceptr)addr),
-                                 log_level);
-    if (UCS_OK != status) {
-        return status;
-    }
 
     UCT_CUDA_IPC_GET_DEVICE(cu_device);
 
-    UCT_CUDADRV_FUNC(cuMemGetAddressRange(&key->d_bptr, &key->b_len,
-                                          (CUdeviceptr)addr),
-                     log_level);
-
-    key->dev_num  = (int) cu_device;
-    key->pid      = getpid();
-    ucs_trace("registered memory:%p..%p length:%lu dev_num:%d",
-              addr, UCS_PTR_BYTE_OFFSET(addr, length), length, (int) cu_device);
-    return UCS_OK;
-}
-
-static ucs_status_t
-uct_cuda_ipc_mem_reg(uct_md_h md, void *address, size_t length,
-                     const uct_md_mem_reg_params_t *params, uct_mem_h *memh_p)
-{
-    uct_cuda_ipc_key_t *key;
-    ucs_status_t status;
-
-    key = ucs_malloc(sizeof(uct_cuda_ipc_key_t), "uct_cuda_ipc_key_t");
-    if (NULL == key) {
+    memh = ucs_malloc(sizeof(uct_cuda_ipc_memh_t), "uct_cuda_ipc_key_t");
+    if (NULL == memh) {
         ucs_error("failed to allocate memory for uct_cuda_ipc_key_t");
         return UCS_ERR_NO_MEMORY;
     }
 
-    status = uct_cuda_ipc_mem_reg_internal(md, address, length, 0, key);
-    if (status != UCS_OK) {
-        ucs_free(key);
-        return status;
-    }
-    *memh_p = key;
+    memh->dev_num = (int) cu_device;
+    memh->pid     = getpid();
+    ucs_list_head_init(&memh->list);
 
+    *memh_p = memh;
     return UCS_OK;
 }
 
 static ucs_status_t
 uct_cuda_ipc_mem_dereg(uct_md_h md,
                        const uct_md_mem_dereg_params_t *params)
 {
+    uct_cuda_ipc_memh_t *memh = params->memh;
+    uct_cuda_ipc_lkey_t *key, *tmp;
+
     UCT_MD_MEM_DEREG_CHECK_PARAMS(params, 0);
 
+    ucs_list_for_each_safe(key, tmp, &memh->list, link) {
+        ucs_free(key);
```

Comment:
do we need to release anything in Cuda?

Reply:
unresolved?

Reply:
AFAIU no, as before, cuIpcGetMemHandle() has no side effects.